### PR TITLE
[5.4] Add deleted callback to Jobs

### PR DESCRIPTION
**Pull request by Belphemur**
When a job end, this callback will be called on it to tell it it's been deleted from the queue.
It's useful when you want to manage jobs that can only have one instance in the queue.

The code is a vulgar copy/paste from the failed callback only with a rename.

**tl;dr**: It would be used to clean up after the job has finished, like a finally in a try/catch clause.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Please update the tests. :)

File: tests/Queue/RedisQueueIntegrationTest.php [hunk 1]
```diff
@@ -303,6 +305,7 @@ public function testSize()
         $this->assertEquals(3, $this->queue->size());
         $job = $this->queue->pop();
         $this->assertEquals(3, $this->queue->size());
+        $job->getContainer()->shouldReceive('make')->once()->with('Illuminate\Queue\CallQueuedHandler')->andReturn($handler = m::mock('StdClass'));
```
Comment (GrahamCampbell) [inline tests/Queue/RedisQueueIntegrationTest.php]: Please use `::class` notation on new code.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Queue/RedisQueueIntegrationTest.php [hunk 2]
```diff
@@ -58,7 +58,8 @@ public function setUp()
         }
 
         $this->queue = new RedisQueue($this->redis);
-        $this->queue->setContainer(m::mock(Container::class));
+        $container   = m::mock(Container::class);
```
Comment (GrahamCampbell) [inline tests/Queue/RedisQueueIntegrationTest.php]: Please don't align assignment operators.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Queue/RedisQueueIntegrationTest.php (hunk 2 above)
Comment (Belphemur) [inline tests/Queue/RedisQueueIntegrationTest.php]: Sorry it's my IDE, usual style for work.
Going to correct it and the last tests that are breaking.

File: tests/Queue/RedisQueueIntegrationTest.php (hunk 1 above)
Comment (Belphemur) [inline tests/Queue/RedisQueueIntegrationTest.php]: If I do that I'm getting

> No matching handler found for Mockery_27_Illuminate_Container_Container::make("Illuminate\Queue\CallQueuedHandler"). Either the method was unexpected or its arguments matched no expected argument list for this method

Which is weird since it's exaclty the same string ...

**Comment by Belphemur** [comment]
I didn't see you moved it on 5.4, I'm going to fix the building since a method changed.

**Comment by Belphemur** [comment]
@GrahamCampbell All StyleCI and test working for 5.4

File: tests/Queue/RedisQueueIntegrationTest.php (hunk 1 above)
Comment (Belphemur) [inline tests/Queue/RedisQueueIntegrationTest.php]: Done and on 5.4 working

File: tests/Queue/QueueBeanstalkdJobTest.php [hunk 3]
```diff
@@ -36,7 +36,8 @@ public function testDeleteRemovesTheJobFromBeanstalkd()
     {
         $job = $this->getJob();
         $job->getPheanstalk()->shouldReceive('delete')->once()->with($job->getPheanstalkJob());
-
+        $job->getPheanstalkJob()->shouldReceive('getData')->once();
+        $job->getContainer()->shouldReceive('make')->once()->with('')->andReturn($handler = m::mock('StdClass'));
         $job->delete();
```
Comment (GrahamCampbell) [inline tests/Queue/QueueBeanstalkdJobTest.php]: Please add a newline before calling this method, like there was before.

**Review by GrahamCampbell — DISMISSED** [review]

File: tests/Queue/QueueBeanstalkdJobTest.php (hunk 3 above)
Comment (Belphemur) [inline tests/Queue/QueueBeanstalkdJobTest.php]: Done, added a line just before the delete

**Comment by taylorotwell** [comment]
So what would you put in this method?

**Comment by Belphemur** [comment]
@taylorotwell In my case, I have a jobScheduler build on top of the normal Job Dispatcher.

I have job that can only be present X times in the queue (for business reason) and need to be sure if an user trigger the scheduling of a job that it doesn't get added to the queue if X limit is reached.

Do count the number of job that have finished, it would be easier to put the decreasing of the job counter in the deleted callback instead of having to put it multiple place in the handle and in the failed callback.

I would say it would be use to clean up after the job has finished, like a finally in a try/catch clause.

**Comment by Belphemur** [comment]
@taylorotwell  @GrahamCampbell 
Just a ping about this pull request about the Jobs:

**tl;dr**: It would be used to clean up after the job has finished, like a finally in a try/catch clause.

**Comment by taylorotwell** [comment]
One concern is this increases the amount of times we serialize the command, which in turn could hit the database with multiple queries every time we delete a job. Especially if the job contains a collection with many items this could incur a performance hit. I think a better option would be to implement an event or something for now maybe.

**Comment by Belphemur** [comment]
@taylorotwell Good point.

Instead of an event, I could implement a memory cache (just an attribute) in CallQueuedHandler to hold the first deserialization of the job (when calling it) and to be reused on failed and deleted.

This way, not only we only do once the deserialization, but on failure, we don't do it anymore.

What do you think?
